### PR TITLE
Fix and check

### PR DESCRIPTION
**Pull request by X-Ryl669**
Here's my take at fixing the bugs and adding the features. I've tested them all and they are working on my units. Please do the same with yours.

## Description of the changes
I've added support for FAN_ONLY, ECONOCOOL, POWERFUL mode. FAN_ONLY is supported natively by ClimateIR but the other 2 aren't so I've used preset for these (it's a mode you can switch to)

I've removed the fan1, fan2 et al, and simplified it (IMHO) to a better system: instead of having the user to specify what Cimate's LOW fan setting is, she will specify what her unit is able to do. So she set a `fan_mode` to `4 levels` or `Quiet + 4 levels` or `5 levels`. The code maps those to Climate's available fan setting. The mapping is not bidirectional, there are more level than what Climate supports, so I've dropped level 2 or 4 when appropriate. It's completely arbitrary, but it work.

Similarly, she sets the `operating_mode` to whatever her unit is able to do (`auto / heat / dry / cool / fan`) and the code will reject request that aren't supported by the unit, to avoid sending the unit in limbo.

### CODE REVIEW

**Comment by X-Ryl669** [comment]
Ok, I think I've resolved all the points you've mentioned in the main thread.
The 4L does use the MIDDLE value (look at the next line, there's a commented test **and** a test for 4L). I've kept the 5L as comment since at least one user reported having it, so he should be able to test it and if it work, I'll re-enable it in a later PR. 

Thanks for your patience.

File: esphome/components/mitsubishi/mitsubishi.cpp [hunk 1]
```diff
@@ -44,39 +49,52 @@ const uint8_t MITSUBISHI_BYTE04 = 0X00;
...
+
+  // Default to only 3 levels in ESPHome even if most unit supports 4. The 3rd level is not used.
   traits.set_supported_fan_modes(
-      {climate::CLIMATE_FAN_LOW, climate::CLIMATE_FAN_MEDIUM, climate::CLIMATE_FAN_HIGH, climate::CLIMATE_FAN_AUTO});
-  if (this->fan_medium_low_)
-    traits.set_supported_custom_fan_modes({CLIMATE_FAN_MEDIUM_LOW});
+      {climate::CLIMATE_FAN_AUTO, climate::CLIMATE_FAN_LOW, climate::CLIMATE_FAN_MEDIUM, climate::CLIMATE_FAN_HIGH});
+  if (this->fan_mode_ == MITSUBISHI_FAN_Q4L)
+    traits.add_supported_fan_mode(climate::CLIMATE_FAN_QUIET);
+  if (/*this->fan_mode_ == MITSUBISHI_FAN_5L ||*/ this->fan_mode_ == MITSUBISHI_FAN_Q4L)
```
Comment (RubyBailey) [inline esphome/components/mitsubishi/mitsubishi.cpp]: Thanks for your help and github guidance. Thought Id try this route for comments, which I think is what you suggested. If not let me know if there's a better way.
I think this check should be for `MITSUBISHI_FAN_4L`, which would then add MIDDLE to 4L. I cannot find a check for 4L anywhere at this time. 
Thanks again for all your work / help / patience :-)

Comment (X-Ryl669) [reply esphome/components/mitsubishi/mitsubishi.cpp]: Ok, I think I've understood your comment this time, sorry ;-)

This change should fix it now. Thanks!

**Comment by X-Ryl669** [comment]
I've modified the code, please review the changes to see if you agree to them. Thank you.

**Comment by RubyBailey** [comment]
When you have a moment would you please look at the esphome-docs PR portion for this code PR at [https://deploy-preview-2353--esphome.netlify.app/components/climate/climate_ir#mitsubishi](url) and let me know if anything needs to be clarified or updated. Per my previous comment I'm unsure how to document the PRESETS. Thanks for your help.

**Comment by X-Ryl669** [comment]
I'm not sure you need to document the presets since it's a runtime option, not a configuration (YAML) option.

No other climate component documents them so it would be awkward to do so, IMHO. One can just select them in HA and it'll work as intuitively expected from their name and what they have on their remote.

I'm ok with the current documentation, it's quite clear and understandable, but I might not be the right one to judge since I've written part of this code.

If I had to be picky and pedantic, I'd add to the `set_x_mode` something like "Select the fan modes as supported on your remote" or "Select the operating modes as supported on your unit" so it's clear that the user must describe her unit's features.

Thank you for updating the documentation!

**Comment by RubyBailey** [comment]
Thanks for your suggestions. Nothing picky or pedantic, I think they help with better documentation. I often say my ODC-ness is just another form of precision :-) 
As you suggested we will leave the presets for now. They show up in HA. That's fine and I'll see if any questions arise. I did some more testing yesterday and they are available in MQTT. All good for now.
Now to see if I can get the PR included in the next ESPHome release. :-)

Again thanks for all you help and patience. You've taught me much about github, esphome and c++.

**Comment by RubyBailey** [comment]
Hello again,
I had an issue with one of my units today, cooling when it should not have been enabled in yaml. I have been troubleshooting the issue and did some more testing with my HVAC's.

I think the set_supported_mode is in conflict with the supports_cool, supports_ heat, supports_dry and supports_fan_only modes. By default all 4 of the latter are True, which then enabled all modes.

mitsubishi.cpp
`if (this->supports_cool_)
    traits.add_supported_mode(climate::CLIMATE_MODE_COOL);
  if (this->supports_heat_)
    traits.add_supported_mode(climate::CLIMATE_MODE_HEAT);

  if (this->supported_mode_ != MITSUBISHI_OP_MODE_AH)
    traits.add_supported_mode(climate::CLIMATE_MODE_COOL);
  if (this->supported_mode_ >= MITSUBISHI_OP_MODE_AHC) {
    traits.add_supported_mode(climate::CLIMATE_MODE_HEAT_COOL);
    traits.add_supported_mode(climate::CLIMATE_MODE_HEAT);
    traits.add_supported_mode(climate::CLIMATE_MODE_COOL);
  }
  if (this->supports_dry_ || this->supported_mode_ >= MITSUBISHI_OP_MODE_ADHC)
    traits.add_supported_mode(climate::CLIMATE_MODE_DRY);
  if (this->supports_fan_only_ || this->supported_mode_ >= MITSUBISHI_OP_MODE_ADFHC)
    traits.add_supported_mode(climate::CLIMATE_MODE_FAN_ONLY);`

Testing it with the following yaml:
`climate:
  - platform: mitsubishi
    name: "Heatpump"
    receiver_id: rcvr
    retain: False
    supports_cool: False
    supports_heat: False`

I can set the unit to cool, heat, dry and fan_only, even though they should be disabled per the yaml.
Cool and heat are enabled by the default set_supported_mode in the above code, and dry and fan_only through the default code settings in mitsubishi.h

I think utilizing the existing options supports_cool, supports_ heat, supports_dry and supports_fan_only to set the acceptable modes would be better and to remove set_supported_mode.  

Yaml example:
`support_cool:  
supports_heat: 
supports_dry: 
supports_fan_only: `

With the default being True, True, False, False per the existing esphome code. 

I know it's another rewrite, though I think in maintains better compatibility with the existing esphome code and removes any code conflicts. I can tackle this one. Would be interested in your thoughts. Thanks.

**Comment by X-Ryl669** [comment]
Please check #26 for an implementation. I'm not 100% sure it's the best long term solution. IMHO, the best solution would be to modify ClimateIR component to support `supports_dry` and `supports_fan_only` but that's not the scope of this PR here. 

So let's implement like I did in the new PR and then if it's accepted upstream, we can propose a change to esphome ClimateIR component in a later PR.

**Pull request merged**